## Import Libraries

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
import xgboost as xgb
import lightgbm as lgb
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns

## Reading Datafile

In [9]:
import pandas as pd
df = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')
print(df.to_string()) 

      age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  ca  thal  target
0      52    1   0       125   212    0        1      168      0      1.0      2   2     3       0
1      53    1   0       140   203    1        0      155      1      3.1      0   0     3       0
2      70    1   0       145   174    0        1      125      1      2.6      0   0     3       0
3      61    1   0       148   203    0        1      161      0      0.0      2   1     3       0
4      62    0   0       138   294    1        1      106      0      1.9      1   3     2       0
5      58    0   0       100   248    0        0      122      0      1.0      1   0     2       1
6      58    1   0       114   318    0        2      140      0      4.4      0   3     1       0
7      55    1   0       160   289    0        0      145      1      0.8      1   1     3       0
8      46    1   0       120   249    0        0      144      0      0.8      2   0     3       0
9      54 

## Exploring Data

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In [11]:
print("Dataset shape:", df.shape)

Dataset shape: (1025, 14)


## Preprocessing and cleaning the data

In [12]:
missing_values = df.isnull().sum()

print("\nMissing values:")
print(missing_values)


Missing values:
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64


In [13]:
df.describe().T

count        mean        std    min    25%    50%    75%    max
age       1025.0   54.434146   9.072290   29.0   48.0   56.0   61.0   77.0
sex       1025.0    0.695610   0.460373    0.0    0.0    1.0    1.0    1.0
cp        1025.0    0.942439   1.029641    0.0    0.0    1.0    2.0    3.0
trestbps  1025.0  131.611707  17.516718   94.0  120.0  130.0  140.0  200.0
chol      1025.0  246.000000  51.592510  126.0  211.0  240.0  275.0  564.0
fbs       1025.0    0.149268   0.356527    0.0    0.0    0.0    0.0    1.0
restecg   1025.0    0.529756   0.527878    0.0    0.0    1.0    1.0    2.0
thalach   1025.0  149.114146  23.005724   71.0  132.0  152.0  166.0  202.0
exang     1025.0    0.336585   0.472772    0.0    0.0    0.0    1.0    1.0
oldpeak   1025.0    1.071512   1.175053    0.0    0.0    0.8    1.8    6.2
slope     1025.0    1.385366   0.617755    0.0    1.0    1.0    2.0    2.0
ca        1025.0    0.754146   1.030798    0.0    0.0    0.0    1.0    4.0
thal      1025.0    2.323902   0.620660    0.0    2.0    2.0    3.0    3.0
target    1025.0    0.513171   0.500070    0.0    0.0    1.0    1.0    1.0

In [14]:
df.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [15]:
df.nunique()

age          41
sex           2
cp            4
trestbps     49
chol        152
fbs           2
restecg       3
thalach      91
exang         2
oldpeak      40
slope         3
ca            5
thal          4
target        2
dtype: int64

In [16]:
df['trestbps'].unique()

array([125, 140, 145, 148, 138, 100, 114, 160, 120, 122, 112, 132, 118,
       128, 124, 106, 104, 135, 130, 136, 180, 129, 150, 178, 146, 117,
       152, 154, 170, 134, 174, 144, 108, 123, 110, 142, 126, 192, 115,
        94, 200, 165, 102, 105, 155, 172, 164, 156, 101], dtype=int64)

# Logistic Regression

In [18]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the Logistic Regression classifier
logistic_model = LogisticRegression(random_state=42)
logistic_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = logistic_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

# Print a classification report for more metrics
report = classification_report(y_test, y_pred)
print('Classification Report:')
print(report)

Accuracy: 0.78
Confusion Matrix:
[[70 32]
 [13 90]]
Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.69      0.76       102
           1       0.74      0.87      0.80       103

    accuracy                           0.78       205
   macro avg       0.79      0.78      0.78       205
weighted avg       0.79      0.78      0.78       205



C:\Users\ssidh\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Accuracy: The model achieves an accuracy of 78%, indicating that it correctly predicts heart attack outcomes for 78% of the cases.

Confusion Matrix:

True Positives (TP): 90
False Positives (FP): 32
True Negatives (TN): 70
False Negatives (FN): 13
This matrix provides a detailed breakdown of correct and incorrect predictions.
Classification Report:

Precision (0): 84% of instances predicted as "0" (no heart attack) were correct.
Precision (1): 74% of instances predicted as "1" (heart attack) were correct.
Recall (0): 69% of actual "0" instances were correctly identified.
Recall (1): 87% of actual "1" instances were correctly identified.
F1-score (0): The balanced F1-score for class "0" is 0.76.
F1-score (1): The balanced F1-score for class "1" is 0.80.
Support: Indicates the number of instances in each class.
Overall, the Logistic Regression model demonstrates effectiveness in providing a reasonably accurate prediction of heart attack outcomes. It emphasizes recall (sensitivity) in identifying actual heart attack cases, resulting in balanced F1-scores, which indicates its potential for risk assessment.

# Decision Tree

In [19]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the Decision Tree classifier
decision_tree_model = DecisionTreeClassifier(random_state=42)
decision_tree_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = decision_tree_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

# Display classification report
report = classification_report(y_test, y_pred)
print('Classification Report:')
print(report)

Accuracy: 0.99
Confusion Matrix:
[[102   0]
 [  3 100]]
Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       102
           1       1.00      0.97      0.99       103

    accuracy                           0.99       205
   macro avg       0.99      0.99      0.99       205
weighted avg       0.99      0.99      0.99       205



Accuracy: The model achieves an exceptional accuracy of 99%, indicating that it correctly predicts heart attack outcomes for 99% of the cases.

Confusion Matrix:

True Positives (TP): 100
False Positives (FP): 0
True Negatives (TN): 102
False Negatives (FN): 3
This confusion matrix reflects the model's remarkable accuracy with very few misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.97, indicating that 97% of instances predicted as not having a heart attack were accurate.
Precision (1): For class "1" (heart attack), the precision is 1.00, meaning that 100% of instances predicted as having a heart attack were correct.
Recall (0): The model correctly identified 100% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 97% of them.
F1-score (0): The F1-score for class "0" is 0.99, reflecting a well-balanced performance.
F1-score (1): The F1-score for class "1" is also 0.99, indicating near perfection.
Support: Indicates the number of instances in each class.
In summary, the Decision Tree model exhibits exceptional effectiveness, achieving near-perfect accuracy, precision, recall, and F1-scores. It excels in accurately identifying cases with and without heart attacks, making it a top choice for this predictive task.

# Ada Boost 

In [24]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the AdaBoost classifier
adaboost_model = AdaBoostClassifier(n_estimators=50, random_state=42)
adaboost_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = adaboost_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

report = classification_report(y_test, y_pred)
print('Classification Report:')
print(report)

Accuracy: 0.88
Confusion Matrix:
[[91 11]
 [14 89]]
Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.89      0.88       102
           1       0.89      0.86      0.88       103

    accuracy                           0.88       205
   macro avg       0.88      0.88      0.88       205
weighted avg       0.88      0.88      0.88       205



Accuracy: The model achieves an accuracy of 88%, indicating that it correctly predicts heart attack outcomes for 88% of the cases.

Confusion Matrix:

True Positives (TP): 89
False Positives (FP): 11
True Negatives (TN): 91
False Negatives (FN): 14
The confusion matrix demonstrates the model's ability to accurately categorize cases with minimal misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.87, indicating that 87% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 0.89, signifying that 89% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 89% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 86% of them.
F1-score (0): The F1-score for class "0" is 0.88, indicating a balanced performance.
F1-score (1): The F1-score for class "1" is also 0.88, reflecting a balanced measure of accuracy.
Support: Indicates the number of instances in each class.

# Support Vector Machine

In [27]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the SVM classifier
svm_model = SVC(kernel='linear', random_state=42)
svm_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = svm_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

report = classification_report(y_test, y_pred)
print('Classification Report:')
print(report)

Accuracy: 0.80
Confusion Matrix:
[[72 30]
 [10 93]]
Classification Report:
              precision    recall  f1-score   support

           0       0.88      0.71      0.78       102
           1       0.76      0.90      0.82       103

    accuracy                           0.80       205
   macro avg       0.82      0.80      0.80       205
weighted avg       0.82      0.80      0.80       205



Accuracy: The model achieves an accuracy of 80%, indicating that it correctly predicts heart attack outcomes for 80% of the cases.

Confusion Matrix:

True Positives (TP): 93
False Positives (FP): 30
True Negatives (TN): 72
False Negatives (FN): 10
The confusion matrix reflects the model's ability to accurately categorize cases while still having some misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.88, indicating that 88% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 0.76, signifying that 76% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 71% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 90% of them.
F1-score (0): The F1-score for class "0" is 0.78, indicating a reasonably balanced performance.
F1-score (1): The F1-score for class "1" is 0.82, reflecting a well-balanced measure of accuracy.
Support: Indicates the number of instances in each class.
In summary, the Support Vector Machine (SVM) model demonstrates a good level of effectiveness in heart attack prediction with an accuracy of 80%. It maintains reasonably balanced precision, recall, and F1-scores, making it a valuable tool for risk assessment.

# K Nearest Neighbor

In [28]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the K-NN classifier (you can adjust the 'n_neighbors' parameter)
knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = knn_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

# Display classification report (includes precision, recall, F1-score, and support)
class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Accuracy: 0.73
Confusion Matrix:
[[74 28]
 [27 76]]
Classification Report:
              precision    recall  f1-score   support

           0       0.73      0.73      0.73       102
           1       0.73      0.74      0.73       103

    accuracy                           0.73       205
   macro avg       0.73      0.73      0.73       205
weighted avg       0.73      0.73      0.73       205



Accuracy: The model achieves an accuracy of 73%, indicating that it correctly predicts heart attack outcomes for 73% of the cases.

Confusion Matrix:

True Positives (TP): 76
False Positives (FP): 28
True Negatives (TN): 74
False Negatives (FN): 27
The confusion matrix reveals the model's ability to categorize cases with a moderate level of correctness but with a relatively higher number of misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.73, indicating that 73% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 0.73, signifying that 73% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 73% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 74% of them.
F1-score (0): The F1-score for class "0" is 0.73, indicating a balanced performance.
F1-score (1): The F1-score for class "1" is also 0.73, reflecting a reasonably balanced measure of accuracy.
Support: Indicates the number of instances in each class.
In summary, the K-Nearest Neighbors (KNN) model demonstrates a moderate level of effectiveness in heart attack prediction with an accuracy of 73%. It maintains balanced precision, recall, and F1-scores but may be less accurate compared to some of the other models mentioned earlier.

# Random Forest Classifier 

In [30]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the Random Forest classifier
random_forest_model = RandomForestClassifier(n_estimators=100, random_state=42)
random_forest_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = random_forest_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Accuracy: 0.99
Confusion Matrix:
[[102   0]
 [  3 100]]
Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       102
           1       1.00      0.97      0.99       103

    accuracy                           0.99       205
   macro avg       0.99      0.99      0.99       205
weighted avg       0.99      0.99      0.99       205



Accuracy: The model achieves an exceptional accuracy of 99%, indicating that it correctly predicts heart attack outcomes for 99% of the cases.

Confusion Matrix:

True Positives (TP): 100
False Positives (FP): 0
True Negatives (TN): 102
False Negatives (FN): 3
The confusion matrix reflects the model's ability to categorize cases with extraordinary accuracy, with very few misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.97, indicating that 97% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 1.00, signifying that 100% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 100% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 97% of them.
F1-score (0): The F1-score for class "0" is 0.99, indicating an exceptional balance of precision and recall.
F1-score (1): The F1-score for class "1" is also 0.99, reflecting near-perfection in accuracy.
Support: Indicates the number of instances in each class.
In summary, the Random Forest model excels in heart attack prediction with an accuracy of 99%. It maintains exceptional precision, recall, and F1-scores, making it a top choice for risk assessment with minimal misclassifications.

# Extra Tree Classifier

In [32]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the Extra Trees classifier
extra_trees_model = ExtraTreesClassifier(n_estimators=100, random_state=42)
extra_trees_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = extra_trees_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Accuracy: 0.99
Confusion Matrix:
[[102   0]
 [  3 100]]
Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       102
           1       1.00      0.97      0.99       103

    accuracy                           0.99       205
   macro avg       0.99      0.99      0.99       205
weighted avg       0.99      0.99      0.99       205



Accuracy: The model achieves an outstanding accuracy of 99%, indicating that it correctly predicts heart attack outcomes for 99% of the cases.

Confusion Matrix:

True Positives (TP): 100
False Positives (FP): 0
True Negatives (TN): 102
False Negatives (FN): 3
The confusion matrix demonstrates the model's ability to categorize cases with remarkable accuracy, with very few misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.97, indicating that 97% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 1.00, signifying that 100% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 100% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 97% of them.
F1-score (0): The F1-score for class "0" is 0.99, reflecting an exceptional balance of precision and recall.
F1-score (1): The F1-score for class "1" is also 0.99, indicating near-perfection in accuracy.
Support: Indicates the number of instances in each class.
In summary, the Extra Tree model excels in heart attack prediction with an accuracy of 99%. It maintains exceptional precision, recall, and F1-scores, demonstrating outstanding performance with minimal misclassifications.

# Gaussian Naive Bayes

In [34]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the Gaussian Naive Bayes classifier
nb_model = GaussianNB()
nb_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = nb_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Accuracy: 0.80
Confusion Matrix:
[[72 30]
 [11 92]]
Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.71      0.78       102
           1       0.75      0.89      0.82       103

    accuracy                           0.80       205
   macro avg       0.81      0.80      0.80       205
weighted avg       0.81      0.80      0.80       205



Accuracy: The model achieves an accuracy of 80%, indicating that it correctly predicts heart attack outcomes for 80% of the cases.

Confusion Matrix:

True Positives (TP): 92
False Positives (FP): 30
True Negatives (TN): 72
False Negatives (FN): 11
The confusion matrix reveals the model's ability to categorize cases with a moderate level of correctness but with a relatively higher number of misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.87, indicating that 87% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 0.75, signifying that 75% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 71% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 89% of them.
F1-score (0): The F1-score for class "0" is 0.78, indicating a reasonably balanced performance.
F1-score (1): The F1-score for class "1" is 0.82, reflecting a well-balanced measure of accuracy.
Support: Indicates the number of instances in each class.
In summary, the Gaussian Naive Bayes (Gaussian NB) model demonstrates a moderate level of effectiveness in heart attack prediction with an accuracy of 80%. It maintains reasonably balanced precision, recall, and F1-scores, making it a valuable tool for risk assessment, although it has more misclassifications compared to the top-performing models.

# Gradient Boosting Classifier

In [35]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the Gradient Boosting Decision Tree classifier
gb_classifier = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)
gb_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = gb_classifier.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

# Print classification report (includes precision, recall, F1-score, and support)
class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Accuracy: 0.93
Confusion Matrix:
[[93  9]
 [ 5 98]]
Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.91      0.93       102
           1       0.92      0.95      0.93       103

    accuracy                           0.93       205
   macro avg       0.93      0.93      0.93       205
weighted avg       0.93      0.93      0.93       205



Accuracy: The model achieves an accuracy of 93%, indicating that it correctly predicts heart attack outcomes for 93% of the cases.

Confusion Matrix:

True Positives (TP): 98
False Positives (FP): 9
True Negatives (TN): 93
False Negatives (FN): 5
The confusion matrix reveals the model's ability to categorize cases with a high degree of correctness and a relatively low number of misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.95, indicating that 95% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 0.92, signifying that 92% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 91% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 95% of them.
F1-score (0): The F1-score for class "0" is 0.93, indicating a well-balanced performance.
F1-score (1): The F1-score for class "1" is also 0.93, reflecting a balanced measure of accuracy.
Support: Indicates the number of instances in each class.
In summary, the Gradient Boosting model performs very well in heart attack prediction with an accuracy of 93%. It maintains balanced precision, recall, and F1-scores, making it a strong choice for risk assessment with a low number of misclassifications.

# XG Boost

In [36]:
pip install xgboost

Note: you may need to restart the kernel to use updated packages.


In [38]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
import xgboost as xgb

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the XGBoost classifier
xgb_model = xgb.XGBClassifier(learning_rate=0.1, max_depth=6, random_state=42)
xgb_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = xgb_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

# Print classification report (includes precision, recall, F1-score, and support)
class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Accuracy: 0.99
Confusion Matrix:
[[102   0]
 [  3 100]]
Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       102
           1       1.00      0.97      0.99       103

    accuracy                           0.99       205
   macro avg       0.99      0.99      0.99       205
weighted avg       0.99      0.99      0.99       205



Accuracy: The model achieves an outstanding accuracy of 99%, indicating that it correctly predicts heart attack outcomes for 99% of the cases.

Confusion Matrix:

True Positives (TP): 100
False Positives (FP): 0
True Negatives (TN): 102
False Negatives (FN): 3
The confusion matrix reflects the model's ability to categorize cases with extraordinary accuracy, with very few misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.97, indicating that 97% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 1.00, signifying that 100% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 100% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 97% of them.
F1-score (0): The F1-score for class "0" is 0.99, indicating an exceptional balance of precision and recall.
F1-score (1): The F1-score for class "1" is also 0.99, reflecting near-perfection in accuracy.
Support: Indicates the number of instances in each class.
In summary, the XGBoost model excels in heart attack prediction with an accuracy of 99%. It maintains exceptional precision, recall, and F1-scores, making it a top choice for risk assessment with minimal misclassifications.

# Light GBM

In [39]:
pip install lightgbm

Note: you may need to restart the kernel to use updated packages.


In [41]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
import lightgbm as lgb

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the LightGBM classifier
lgb_model = lgb.LGBMClassifier(num_leaves=31, learning_rate=0.05, n_estimators=1000, random_state=42)
lgb_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = lgb_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

# Print classification report (includes precision, recall, F1-score, and support)
class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

[LightGBM] [Info] Number of positive: 423, number of negative: 397
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001246 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 363
[LightGBM] [Info] Number of data points in the train set: 820, number of used features: 13
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.515854 -> initscore=0.063436
[LightGBM] [Info] Start training from score 0.063436
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[

[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No f

Accuracy: 0.99
Confusion Matrix:
[[102   0]
 [  3 100]]
Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       102
           1       1.00      0.97      0.99       103

    accuracy                           0.99       205
   macro avg       0.99      0.99      0.99       205
weighted avg       0.99      0.99      0.99       205



Accuracy: The model achieves an outstanding accuracy of 99%, indicating that it correctly predicts heart attack outcomes for 99% of the cases.

Confusion Matrix:

True Positives (TP): 100
False Positives (FP): 0
True Negatives (TN): 102
False Negatives (FN): 3
The confusion matrix reflects the model's ability to categorize cases with extraordinary accuracy, with very few misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.97, indicating that 97% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 1.00, signifying that 100% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 100% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 97% of them.
F1-score (0): The F1-score for class "0" is 0.99, indicating an exceptional balance of precision and recall.
F1-score (1): The F1-score for class "1" is also 0.99, reflecting near-perfection in accuracy.
Support: Indicates the number of instances in each class.
In summary, the Light GBM model excels in heart attack prediction with an accuracy of 99%. It maintains exceptional precision, recall, and F1-scores, making it a top choice for risk assessment with minimal misclassifications.

# Cat Boost

In [42]:
pip install catboost

Note: you may need to restart the kernel to use updated packages.


In [4]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from catboost import CatBoostClassifier

# Load your dataset (replace 'your_dataset.csv' with the actual path to your dataset)
data = pd.read_csv(r'C:\Users\ssidh\OneDrive\Documents\ML PROJECT SEM V\Datasets\heart relate.csv')

# Split the data into features (X) and the target variable (y)
X = data.drop(columns=['target'])
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the CatBoost classifier
catboost_model = CatBoostClassifier(iterations=1000, learning_rate=0.1, depth=6, random_state=42, cat_features=['sex', 'cp', 'fbs', 'restecg', 'exang', 'slope', 'ca', 'thal'])
catboost_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = catboost_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

# Display classification report with precision, recall, F1-score, and support
report = classification_report(y_test, y_pred)
print('Classification Report:')
print(report)

0:	learn: 0.5897543	total: 280ms	remaining: 4m 39s
1:	learn: 0.5190448	total: 338ms	remaining: 2m 48s
2:	learn: 0.4682426	total: 388ms	remaining: 2m 8s
3:	learn: 0.4177078	total: 434ms	remaining: 1m 48s
4:	learn: 0.3922428	total: 476ms	remaining: 1m 34s
5:	learn: 0.3626152	total: 534ms	remaining: 1m 28s
6:	learn: 0.3387046	total: 582ms	remaining: 1m 22s
7:	learn: 0.3146256	total: 625ms	remaining: 1m 17s
8:	learn: 0.2880500	total: 669ms	remaining: 1m 13s
9:	learn: 0.2650150	total: 733ms	remaining: 1m 12s
10:	learn: 0.2513164	total: 807ms	remaining: 1m 12s
11:	learn: 0.2453415	total: 853ms	remaining: 1m 10s
12:	learn: 0.2376739	total: 897ms	remaining: 1m 8s
13:	learn: 0.2339854	total: 956ms	remaining: 1m 7s
14:	learn: 0.2291424	total: 1.04s	remaining: 1m 8s
15:	learn: 0.2201801	total: 1.12s	remaining: 1m 8s
16:	learn: 0.2143565	total: 1.17s	remaining: 1m 7s
17:	learn: 0.2036412	total: 1.21s	remaining: 1m 6s
18:	learn: 0.1936985	total: 1.26s	remaining: 1m 5s
19:	learn: 0.1894268	total: 1.

161:	learn: 0.0236763	total: 8.17s	remaining: 42.3s
162:	learn: 0.0235254	total: 8.22s	remaining: 42.2s
163:	learn: 0.0234286	total: 8.26s	remaining: 42.1s
164:	learn: 0.0234220	total: 8.3s	remaining: 42s
165:	learn: 0.0232497	total: 8.35s	remaining: 41.9s
166:	learn: 0.0229628	total: 8.39s	remaining: 41.9s
167:	learn: 0.0229628	total: 8.44s	remaining: 41.8s
168:	learn: 0.0229065	total: 8.48s	remaining: 41.7s
169:	learn: 0.0228611	total: 8.52s	remaining: 41.6s
170:	learn: 0.0227972	total: 8.56s	remaining: 41.5s
171:	learn: 0.0224545	total: 8.61s	remaining: 41.4s
172:	learn: 0.0223558	total: 8.65s	remaining: 41.4s
173:	learn: 0.0221974	total: 8.7s	remaining: 41.3s
174:	learn: 0.0220910	total: 8.75s	remaining: 41.2s
175:	learn: 0.0220865	total: 8.81s	remaining: 41.2s
176:	learn: 0.0216993	total: 8.89s	remaining: 41.4s
177:	learn: 0.0214547	total: 8.95s	remaining: 41.3s
178:	learn: 0.0213932	total: 9s	remaining: 41.3s
179:	learn: 0.0212361	total: 9.05s	remaining: 41.2s
180:	learn: 0.02096

321:	learn: 0.0095260	total: 16.1s	remaining: 33.8s
322:	learn: 0.0094750	total: 16.1s	remaining: 33.8s
323:	learn: 0.0094749	total: 16.2s	remaining: 33.8s
324:	learn: 0.0094294	total: 16.2s	remaining: 33.7s
325:	learn: 0.0094292	total: 16.3s	remaining: 33.6s
326:	learn: 0.0093982	total: 16.3s	remaining: 33.6s
327:	learn: 0.0093081	total: 16.4s	remaining: 33.5s
328:	learn: 0.0093081	total: 16.4s	remaining: 33.5s
329:	learn: 0.0092504	total: 16.5s	remaining: 33.4s
330:	learn: 0.0092501	total: 16.5s	remaining: 33.3s
331:	learn: 0.0091992	total: 16.5s	remaining: 33.3s
332:	learn: 0.0091992	total: 16.6s	remaining: 33.2s
333:	learn: 0.0091823	total: 16.6s	remaining: 33.2s
334:	learn: 0.0091823	total: 16.7s	remaining: 33.1s
335:	learn: 0.0091569	total: 16.7s	remaining: 33.1s
336:	learn: 0.0091187	total: 16.8s	remaining: 33s
337:	learn: 0.0090605	total: 16.8s	remaining: 33s
338:	learn: 0.0090092	total: 16.9s	remaining: 32.9s
339:	learn: 0.0090092	total: 16.9s	remaining: 32.9s
340:	learn: 0.00

480:	learn: 0.0063044	total: 23.7s	remaining: 25.6s
481:	learn: 0.0063044	total: 23.8s	remaining: 25.6s
482:	learn: 0.0063044	total: 23.8s	remaining: 25.5s
483:	learn: 0.0063044	total: 23.9s	remaining: 25.5s
484:	learn: 0.0063042	total: 23.9s	remaining: 25.4s
485:	learn: 0.0063042	total: 24s	remaining: 25.3s
486:	learn: 0.0063042	total: 24s	remaining: 25.3s
487:	learn: 0.0063042	total: 24.1s	remaining: 25.3s
488:	learn: 0.0062872	total: 24.1s	remaining: 25.2s
489:	learn: 0.0062872	total: 24.2s	remaining: 25.2s
490:	learn: 0.0062872	total: 24.2s	remaining: 25.1s
491:	learn: 0.0062872	total: 24.3s	remaining: 25s
492:	learn: 0.0062872	total: 24.3s	remaining: 25s
493:	learn: 0.0062871	total: 24.3s	remaining: 24.9s
494:	learn: 0.0062717	total: 24.4s	remaining: 24.9s
495:	learn: 0.0062716	total: 24.4s	remaining: 24.8s
496:	learn: 0.0062716	total: 24.5s	remaining: 24.8s
497:	learn: 0.0062716	total: 24.5s	remaining: 24.7s
498:	learn: 0.0062716	total: 24.6s	remaining: 24.7s
499:	learn: 0.006271

639:	learn: 0.0059540	total: 31.3s	remaining: 17.6s
640:	learn: 0.0059539	total: 31.3s	remaining: 17.6s
641:	learn: 0.0059539	total: 31.4s	remaining: 17.5s
642:	learn: 0.0059539	total: 31.4s	remaining: 17.4s
643:	learn: 0.0059539	total: 31.5s	remaining: 17.4s
644:	learn: 0.0059119	total: 31.6s	remaining: 17.4s
645:	learn: 0.0059119	total: 31.6s	remaining: 17.3s
646:	learn: 0.0059119	total: 31.7s	remaining: 17.3s
647:	learn: 0.0059084	total: 31.7s	remaining: 17.2s
648:	learn: 0.0059084	total: 31.7s	remaining: 17.2s
649:	learn: 0.0059084	total: 31.8s	remaining: 17.1s
650:	learn: 0.0059083	total: 31.8s	remaining: 17.1s
651:	learn: 0.0059082	total: 31.9s	remaining: 17s
652:	learn: 0.0059082	total: 31.9s	remaining: 17s
653:	learn: 0.0058846	total: 32s	remaining: 16.9s
654:	learn: 0.0058838	total: 32s	remaining: 16.9s
655:	learn: 0.0058838	total: 32.1s	remaining: 16.8s
656:	learn: 0.0058837	total: 32.1s	remaining: 16.8s
657:	learn: 0.0058827	total: 32.1s	remaining: 16.7s
658:	learn: 0.005882

799:	learn: 0.0055649	total: 39s	remaining: 9.75s
800:	learn: 0.0055649	total: 39.1s	remaining: 9.71s
801:	learn: 0.0055649	total: 39.1s	remaining: 9.65s
802:	learn: 0.0055649	total: 39.2s	remaining: 9.61s
803:	learn: 0.0055649	total: 39.2s	remaining: 9.55s
804:	learn: 0.0055649	total: 39.2s	remaining: 9.5s
805:	learn: 0.0055649	total: 39.3s	remaining: 9.45s
806:	learn: 0.0055649	total: 39.3s	remaining: 9.4s
807:	learn: 0.0055649	total: 39.4s	remaining: 9.36s
808:	learn: 0.0055649	total: 39.4s	remaining: 9.31s
809:	learn: 0.0055649	total: 39.5s	remaining: 9.26s
810:	learn: 0.0055649	total: 39.5s	remaining: 9.21s
811:	learn: 0.0055649	total: 39.6s	remaining: 9.16s
812:	learn: 0.0055649	total: 39.6s	remaining: 9.11s
813:	learn: 0.0055649	total: 39.7s	remaining: 9.06s
814:	learn: 0.0055649	total: 39.7s	remaining: 9.01s
815:	learn: 0.0055648	total: 39.8s	remaining: 8.96s
816:	learn: 0.0055648	total: 39.8s	remaining: 8.91s
817:	learn: 0.0055648	total: 39.8s	remaining: 8.86s
818:	learn: 0.00

959:	learn: 0.0053749	total: 46.7s	remaining: 1.95s
960:	learn: 0.0053749	total: 46.8s	remaining: 1.9s
961:	learn: 0.0053749	total: 46.8s	remaining: 1.85s
962:	learn: 0.0053748	total: 46.9s	remaining: 1.8s
963:	learn: 0.0053748	total: 46.9s	remaining: 1.75s
964:	learn: 0.0053748	total: 47s	remaining: 1.7s
965:	learn: 0.0053748	total: 47s	remaining: 1.66s
966:	learn: 0.0053748	total: 47.1s	remaining: 1.61s
967:	learn: 0.0053747	total: 47.1s	remaining: 1.56s
968:	learn: 0.0053747	total: 47.2s	remaining: 1.51s
969:	learn: 0.0053747	total: 47.2s	remaining: 1.46s
970:	learn: 0.0053747	total: 47.2s	remaining: 1.41s
971:	learn: 0.0053747	total: 47.3s	remaining: 1.36s
972:	learn: 0.0053747	total: 47.3s	remaining: 1.31s
973:	learn: 0.0053747	total: 47.3s	remaining: 1.26s
974:	learn: 0.0053747	total: 47.4s	remaining: 1.21s
975:	learn: 0.0053747	total: 47.4s	remaining: 1.17s
976:	learn: 0.0053746	total: 47.5s	remaining: 1.12s
977:	learn: 0.0053746	total: 47.5s	remaining: 1.07s
978:	learn: 0.00537

Accuracy: The model achieves an accuracy of 98%, indicating that it correctly predicts heart attack outcomes for 98% of the cases.

Confusion Matrix:

True Positives (TP): 99
False Positives (FP): 0
True Negatives (TN): 102
False Negatives (FN): 4
The confusion matrix reflects the model's ability to categorize cases with very high accuracy and very few misclassifications.
Classification Report:

Precision (0): The precision for class "0" (no heart attack) is 0.96, indicating that 96% of instances predicted as not having a heart attack were correct.
Precision (1): For class "1" (heart attack), the precision is 1.00, signifying that 100% of instances predicted as having a heart attack were accurate.
Recall (0): The model correctly identified 100% of the actual cases with no heart attack (class "0").
Recall (1): For cases with a heart attack (class "1"), the model accurately identified 96% of them.
F1-score (0): The F1-score for class "0" is 0.98, indicating a balanced performance.
F1-score (1): The F1-score for class "1" is also 0.98, reflecting a well-balanced measure of accuracy.
Support: Indicates the number of instances in each class.
In summary, the CatBoost model performs very well in heart attack prediction with an accuracy of 98%. It maintains balanced precision, recall, and F1-scores, making it a strong choice for risk assessment with very few misclassifications.

# Applying models for prediciton

In [72]:
print("Please provide the following information:")

age = float(input("\nAge: "))
sex = int(input("Sex (0 for female, 1 for male): "))
cp = int(input("Chest pain type (0-3): "))
trestbps = float(input("Resting blood pressure (mm Hg): "))
chol = float(input("Serum cholesterol level (mg/dL): "))
fbs = int(input("Fasting blood sugar > 120 mg/dL (0 for false, 1 for true): "))
restecg = int(input("Resting electrocardiographic results (0-2): "))
thalach = float(input("Maximum heart rate achieved during exercise: "))
exang = int(input("Exercise-induced angina (0 for no, 1 for yes): "))
oldpeak = float(input("ST depression induced by exercise relative to rest: "))
slope = int(input("Slope of the peak exercise ST segment (0-2): "))
ca = int(input("Number of major vessels colored by fluoroscopy (0-3): "))
thal = int(input("Thallium stress test result (0-2): "))

prediction = catboost_model.predict([[age, sex, cp,  trestbps ,chol, fbs, restecg , thalach, exang , oldpeak, slope, ca, thal]])
print("\nRISK RATE : {}".format(prediction))

if prediction == 0:
    print('You are not at risk of a heart attack')
else:
    print('You are at risk of a heart attack.')

Please provide the following information:

Age: 55
Sex (0 for female, 1 for male): 1
Chest pain type (0-3): 1
Resting blood pressure (mm Hg): 130
Serum cholesterol level (mg/dL): 220
Fasting blood sugar > 120 mg/dL (0 for false, 1 for true): 0
Resting electrocardiographic results (0-2): 0
Maximum heart rate achieved during exercise: 160
Exercise-induced angina (0 for no, 1 for yes): 0
ST depression induced by exercise relative to rest: 1.5
Slope of the peak exercise ST segment (0-2): 2
Number of major vessels colored by fluoroscopy (0-3): 0
Thallium stress test result (0-2): 2

RISK RATE : [1]
You are at risk of a heart attack.


Age: The age of the individual in years. Age is a significant factor in heart disease, as the risk of heart attacks generally increases with age.

Sex: Gender of the individual (0 for female, 1 for male). Gender can be a relevant factor, as men tend to have a higher risk of heart attacks than women.

Chest Pain Type (CP): This feature represents the type of chest pain experienced by the individual. It is typically classified into four categories (0-3), with higher values indicating more severe chest pain. The type of chest pain can provide insights into the likelihood of a heart attack.

Resting Blood Pressure (trestbps): This is the resting blood pressure of the individual in millimeters of mercury (mm Hg). High blood pressure (hypertension) is a known risk factor for heart disease.

Serum Cholesterol Level (chol): The serum cholesterol level in milligrams per deciliter (mg/dL). High levels of cholesterol in the blood can contribute to the development of atherosclerosis, which increases the risk of heart attacks.

Fasting Blood Sugar (fbs): Indicates whether the individual has fasting blood sugar greater than 120 mg/dL (0 for false, 1 for true). Elevated blood sugar can be associated with diabetes, which is a risk factor for heart disease.

Resting Electrocardiographic Results (restecg): Represents the results of the resting electrocardiogram (ECG or EKG). It is typically classified into three categories (0-2) and can reveal abnormal heart rhythms or signs of heart damage.

Maximum Heart Rate Achieved (thalach): The maximum heart rate achieved during exercise. This can be an important indicator of cardiovascular fitness and can help assess heart health.

Exercise-Induced Angina (exang): Indicates whether the individual experiences exercise-induced angina (0 for no, 1 for yes). Angina is chest pain or discomfort that occurs when the heart muscle doesn't get enough oxygen-rich blood. It can be a symptom of heart disease.

ST Depression Induced by Exercise (oldpeak): Represents the ST segment depression induced by exercise relative to rest. This can indicate myocardial ischemia, which is a lack of blood flow to the heart muscle.

Slope of the Peak Exercise ST Segment (slope): Describes the slope of the peak exercise ST segment (0-2). The shape of this segment can provide insights into the heart's response to exercise.

Number of Major Vessels Colored by Fluoroscopy (ca): Represents the number of major blood vessels colored by fluoroscopy (0-3). It is a measure of coronary artery disease, which can restrict blood flow to the heart.

Thallium Stress Test Result (thal): Indicates the result of the thallium stress test (0-2). This test can show how well blood flows into the heart muscle, providing information about heart health.

These features are used to build predictive models that assess an individual's risk of experiencing a heart attack based on their demographic and clinical characteristics. 